### PR TITLE
Gossipsub: Use sets instead of vectors inside behaviour state

### PR DESCRIPTION
**Pull request by rklaehn**
To completely eliminate the possibility of having duplicate peers in these sets

### CODE REVIEW

File: protocols/gossipsub/src/behaviour.rs [hunk 1]
```diff
@@ -771,7 +766,9 @@ impl Gossipsub {
                     }
                 }
             }
-            peers.retain(|peer| to_remove_peers.contains(&peer));
```
Comment (rklaehn) [inline protocols/gossipsub/src/behaviour.rs:774]: This is different than in the old code, but I think the old code has the condition flipped.

Comment (AgeManning) [reply protocols/gossipsub/src/behaviour.rs:774]: Yes true. Nice catch

**Comment by AgeManning** [comment]
This a nice addition. There are some extra checks (for duplicates in the Vectors) that we no longer need and can be removed. 

I'm going to try and finish up fixing all these bugs and hopefully address your issue in #1583. To this end, I've merged these into there.

**Comment by rklaehn** [comment]
So the signing PR is basically your development branch? I might have a few DRY refactoring improvements that I will PR against that one then.

Anyway, I am going to close this PR. Thanks for the inclusion.

**Pull request closed without merging**